### ISSUE TITLE
New JSON-based AutoDiscovery 

### ISSUE THREAD

**Issue by TCB13** [gronke/email-autodiscover#4]
Hello,

I've noticed that the latest Outlook 2016 from Office 365 uses a new auto configuration format, have a look at this from my Apache:

` [06/Apr/2018:04:26:15 +0200] "GET /autodiscover/autodiscover.json?Email=user%40exampledomain.org&Protocol=ActiveSync&RedirectCount=1 HTTP/1.1" 404 4020 "-" "OutlookMobileCloudService-Autodetect/1.0.0"`

Are there any ideas on supporting this newer JSON format as well? Thank you.

**Comment by Billyboylindien** [comment]
Same on my side.
It use activesync protocol.
Impossible to find any documentation :'(
If someone have I take ;)

My searches:
https://github.com/mailcow/mailcow-dockerized/blob/master/data/web/autodiscover-json.php

**Comment by gronke** [comment]
> I've noticed that the latest Outlook 2016 from Office 365 uses a new auto configuration format
@TCB13 do you know if the client sends a user agent string or how we can differentiate which client is connecting?

> Are there any ideas on supporting this newer JSON format as well? Thank you.
It invites for a conditional that differentiates which format to present.

> Impossible to find any documentation :'(
Usually looking at other ISPs configuration is an appropriate way to gain such information.

**Comment by TCB13** [comment]
@gronke I think the link by @Billyboylindien answers that question https://github.com/mailcow/mailcow-dockerized/blob/master/data/web/autodiscover-json.php

Looks like the `Protocol` field in the GET request can be either `ActiveSync` or `AutodiscoverV1`.

**Comment by Gorjan-Relevantmobile** [comment]
Hi Guys, could you please tell me if some one has implemented this?

**Comment by Deepakbaghel99** [comment]
Autodiscover.json is the new Autodiscover (Autodiscover V2) service that is used by Office 365 clients to discover the location of the mailbox to connect too. 
3rd party clients are advised to used Graph.microsoft.com. Behind the scene Graph would do the Autodiscover V2 to discover the protocol URL required to connect to the requested mailbox and then connect to it to get the requested content.
https://channel9.msdn.com/Events/Ignite/2016/BRK3045 is the link to the Ignite talk talking (24.30 mins) about Autodiscover V2 in Graph context.

**Comment by TCB13** [comment]
I still can't get arond this. Outlook doesn't even seem to behave the same from domain to domain.

**Comment by FoZo** [comment]
Hi,
it looks like the new otulook just don't support autodiscovery for non MS services.
After few days digging in the MS documentation I had found this:
https://docs.microsoft.com/en-us/openspecs/exchange_server_protocols/ms-oxdscli/b79961b5-ff44-47bf-b136-b5613c45f18b#Appendix_A_36

**Comment by kartikayx** [comment]
Can someone point me to a document or something ? I need to implement auto discovery for my native mobile app using graph i don't find any resource related to this.

**Comment by kartikayx** [comment]
> Autodiscover.json is the new Autodiscover (Autodiscover V2) service that is used by Office 365 clients to discover the location of the mailbox to connect too.
> 3rd party clients are advised to used Graph.microsoft.com. Behind the scene Graph would do the Autodiscover V2 to discover the protocol URL required to connect to the requested mailbox and then connect to it to get the requested content.
> https://channel9.msdn.com/Events/Ignite/2016/BRK3045 is the link to the Ignite talk talking (24.30 mins) about Autodiscover V2 in Graph context.

@Deepakbaghel99 How we can use graph ? graph will require the auth tokens right ? which api of graph we need to call ?

**Comment by TCB13** [comment]
> 3rd party clients are advised to used Graph.microsoft.com. Behind the scene Graph would do the Autodiscover V2 to discover the protocol URL required to connect to the requested mailbox and then connect to it to get the requested content.

Yeah I've noticed that Microsoft probes my servers for the v2 protocol however they don't seem to pass the settings to Outlook after that... I think this is anti competitive behavior - making sure only O365 email works properly with Outlook. 
Before someone says my v2 implementation is broken just consider for a second that it works just fine with older versions of Outlook but its broken with the newer versions that use microsoft's graph servers.

**Comment by chris001** [comment]
I'm surprised and don't believe MS is still acting anti competitive in 2020.  Example, they recently integrated linux VMs into windows server (afaik)!  I bet this AutodiscoverV2 is caused by human error of missing documentation on how to use their Graph API and it should appear soon, as long we ask MS for it, and/or search in MS Tech Net for it.

**Comment by Trimilur** [comment]
Are there any news here? I am also unable to get autodiscover with imap working in outlook with o365.

**Comment by christophsanz** [comment]
I am having the same issue. It is weird how if you use the old "Mail (Microsoft Outlook) (32-Bit)" via the Control Panel everything works... this is just frustrating.
Why is automatic email configuration still that hard in 2021, why does not everyone use RFC 6186 ??

**Comment by m3nu** [comment]
There is also a registry flag that can change the default dialog in Outlook back to the “advanced version”.
